### ISSUE TITLE
Fetch all workspaces on every pull-sync and detect missing/added ones

### ISSUE THREAD

**Issue by amulware** [toggl/mobileapp#1336]
There is really no issue with pulling all workspaces (almost always 1-2, very rarely more than 5) of a user on every pull sync.

When persisting the full list we should compare it to the list of local workspaces, and where they differ resolve conflicts accordingly, IN THIS ORDER:

If we lose access (don't confuse this with the existence of a local WS that was not synced yet!):
- [ ] remove all synced TEs from the workspace
- [ ] remove all synced projects/clients/tags that are not referred to by any TE (clients are referred through the projects and their check is maybe a special case)
- [ ] #1337 mark all remaining (i.e. unsynced) entities especially, so that they cannot show up in auto-completion, suggestions. TEs like that should not be able to be continued either

If we gain access to a new WS:
- [ ] add that workspace to the DB
- [ ] make sure that if we had access to this WS before, and there is unsynced data, all that can sync again (we may need to undo whatever we did to them in #1377), and that data is merged correctly when pulling
- [ ] then reset all ?since dates to null
- [ ] go to the top of the pull sync (which will automatically fetch all data for those workspaces, and also for other workspaces, but that's something to optimise later if we need to) - this prevents the stuff from below from running until after the full pull

If the user has no workspaces left:
- [ ] check this by looking if the database has any non-ghost (#1337) workspaces, it does not matter if they are synced or not
- [ ] throw an exception and halt syncing if this happens - the exception will cause a prompt making the user create or retry sync (and the retry will cause another exception if nothing has changed; if it completes however without throwing that exception, the prompt can be closed)

Further:
- [ ] on app start/activation(every X minutes? - depending on the cost of this operation) all 'invalid' entities that are no longer referred to by any time entries can be removed (alternatively, when TEs are edited/deleted, but that might be more complicated?)

See also: #1334

**Comment by simonrozsival** [comment]
Do I understand it correctly that we are gonna keep the workspaces to which the user loses access (which have some unsynced data)  in a somewhat similar state to "ghost projects" - it will be a "ghost workspace".

I am not sure how will we mark the unsynced time entries/projects/clients which belong to this workspace - will we make them "unsynced ghosts"? or will we add another property to the database which will that they are inaccessible (something that I "solved" for ghost projects by setting `Active` to `false`).

Gaining access to a workspace back means that we fetch a workspace which is locally marked as inaccessible - now I guess that we should first update the workspace data in the DB and only then start making the related entities accessible, right?

**Comment by simonrozsival** [comment]
We should wait for #2331 to be closed first before we start implementing this.

**Comment by amulware** [comment]
I think the answer is (mostly) yes to all the questions above.

We'll ghost the workspace, and all clients, projects, tasks, tags. Time entries we do not want to ghost, since they need to show up in the log - unless we just decide that we'll ignore the ghost-flag for log-showing purposes.

Whether we need another property or not I'm not sure right now, that needs some thinking. Maybe another/the same sync state that the projects have is fine? Except we need to tell synced/unsynced ghost TEs apart, unless we delete the synced ones right away.

Gaining access back is... annoying.

Proposal: if we EVER detect that we get a workspace we didn't have before, let's trigger a full sync that ignores SINCE dates, which will cause it to download all data for new workspaces (and all data for other workspaces, but we can't really filter by /me and /workspace scope at the same time I think) - this should override all local ghost data and remove ghost-status. Unless the local data has been changed more recently in the app maybe? What do we do then? It seems we might wanna remove ghost, but still keep the local changes?

**Comment by simonrozsival** [comment]
Will the user be able to make changes to the local "ghost" data? Please say no.

**Comment by simonrozsival** [comment]
(Right now I have time entries in mind - they should be visible in the log, so by default user could continue/delete/edit them.)

**Comment by amulware** [comment]
Actually, yes. The user should be able to make changes to fix things up. For example, if they have an unsynced time entry in a workspace they got removed from, they should be able to change it to a different workspace. Similar with other entities, once we add editing for them. Or at least, this should be possible if the entry was never synced. Not sure about those where only some changes are pending :/

**Comment by amulware** [comment]
The user should be able to delete the entries too, which helps them fix remaining unsynced data. Continuing though, should be forbidden.

**Comment by simonrozsival** [comment]
Oh this issue will be much more complicated. Could we somehow split this into several smaller issues?

And I think we need a different attribute for our entities other than the `SyncStatus.RefetchingNeeded` - if user can change the data, it must not be marked this way. That sync staus should be just for placeholder entities (like for real ghosts, not for somehow "inaccessible" entities).

**Comment by amulware** [comment]
Can you come up with a proposal for what that property should do exactly, and how all the different cases are covered and handled?

And yeah, sure, we can split this up. I'd say, let's do that as soon as we agree on the proposal.

**Comment by amulware** [comment]
(Note that the above discussion is very much in the scope of #1337 as well.)

**Comment by amulware** [comment]
Superseded.
